pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+                                              exptime=self.exptime)
         pypeitImage.rawheadlist = self.headarr
         pypeitImage.process_steps = [key for key in self.steps.keys() if self.steps[key]]
 
         # Mask(s)
-        if par['mask_cr']:
+        if self.par['mask_cr']:
             pypeitImage.build_crmask(self.par)
-        #
         nonlinear_counts = self.spectrograph.nonlinear_counts(self.detector,
                                                               apply_gain=self.par['apply_gain'])
-        # Build
         pypeitImage.build_mask(saturation=nonlinear_counts)
 
         # Return
         return pypeitImage
 
-    def flatten(self, pixel_flat, illum_flat=None, bpm=None, force=False):
+    def spatial_flexure_shift(self, slits, force=False):
         """
-        Flat field the proc image
+        Calculate a spatial shift in the edge traces due to flexure.
 
-        Wrapper to flat.flatfield
+        This is a simple wrapper for
+        :func:`~pypeit.core.flexure.spat_flexure_shift`.
 
         Args:
-            pixel_flat (`numpy.ndarray`_):
-                Pixel flat image
-            illum_flat (`numpy.ndarray`_, optional):
-                Illumination flat image
-            bpm (`numpy.ndarray`_, optional):
-                Bad pixel mask image;  if provided, over-rides internal one
-            force (bool, optional):
-                Force the processing even if the image was already processed
+            slits (:class:`~pypeit.slittrace.SlitTraceSet`, optional):
+                Slit edge traces
+            force (:obj:`bool`, optional):
+                Force the image to be field flattened, even if the step log
+                (:attr:`steps`) indicates that it already has been.
 
         """
         step = inspect.stack()[0][3]
-        # Check if already trimmed
-        if self.steps[step] and (not force):
-            msgs.warn("Image was already flat fielded.  Returning the current image")
-            return self.image.copy()
-        # Check
-        if pixel_flat is None:
-            msgs.error("You requested flattening but provided no pixel flat!")
-        # BPM
-        if bpm is None:
-            bpm = self.bpm
-        # Do it
-        self.image = flat.flatfield(self.image, pixel_flat, bpm, illum_flat=illum_flat)
+        if self.steps[step] and not force:
+            # Already field flattened
+            msgs.warn('Spatial flexure shift already calculated.')
+            return
+        self.spat_flexure_shift = flexure.spat_flexure_shift(self.image, slits)
         self.steps[step] = True
 
-    def orient(self, force=False):
+    def flatten(self, flatimages, slits=None, force=False, debug=False):
         """
-        Orient the image in the PypeIt format with spectra running blue (down)
-        to red (up).
+        Field flatten the processed image.
+
+        This method calculates a slit-illumination correction and a spectral
```

Comment:
It does not actually calculate these quantities. The slit illumination is tied to the slits, and the slits can move because of flexure. So what this does is evaluate the slit illumination correction fits and then applies to them to the images (possibly) taking the flexure into account. I would not say that it "calculates slit illumination or spectral illumination corrections" since that is done in the FlatField code.

Reply:
Sounds good.  Changed to: "This method uses the results of the flat-field modeling code (see :class:`~pypeit.flatfield.FlatField`) and any measured spatial shift due to flexure to construct slit-illumination, spectral response, and pixel-to-pixel response corrections, and multiplicatively removes them from the current image."  Happy to adjust more, if you want.